### PR TITLE
Fix weight loading in RetinaNet and introduce unit tests to ensure weight loading works as intended

### PR DESCRIPTION
**Pull request by LukeWood**
Before this PR there was a bug where Keras failed to track the weights of some of the weights.  This was caused by the fact that weight tracking is handled in `keras.Model` during attribute assignment by checking if the assigned member is a keras layer.  If you use a closure style "layer" by assignment and then use the assigned layer in a custom model, you will not track the weights.

This PR fixes that bug - `load_weights(path)` case and `set_weights(weights)` both work now.

Fixes:
- https://github.com/keras-team/keras-cv/issues/767
- https://github.com/keras-team/keras-cv/issues/766

### CODE REVIEW

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py [hunk 1]
```diff
@@ -12,11 +12,13 @@
...
 # limitations under the License.
 
+import tensorflow as tf
 from tensorflow.keras import initializers
 from tensorflow.keras import layers
 
 
-def PredictionHead(output_filters, bias_initializer):
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class PredictionHead(layers.Layer):
```
Comment (fchollet) [inline keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:21]: So if you use closure-style functions, the layers used are never assigned anywhere on the subclassed models. For them to be tracked they need to be assigned at some point.

For this reason closure-style layers are really for Functional models (where the tracking happens at model construction time). They should not be assigned to a subclassed model. Maybe we could find a way to address this (e.g. by tracking layers called within the `__call__` scope, to be investigated).

**Review by fchollet — APPROVED** [review]
LGTM, thanks for the PR

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py (hunk 1 above)
Comment (qlzh727) [reply keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:21]: Agreed. I think we should use the subclass layer as the standard. Having the __call__ to track those layer might means more magic in the keras base layer. There will be more corner case when save/load happens (what's the attribute name to serialize and desalinize those layers.)

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py [hunk 2]
```diff
@@ -27,30 +29,42 @@ def PredictionHead(output_filters, bias_initializer):
...
+    def __init__(self, output_filters, bias_initializer, **kwargs):
+        super().__init__(**kwargs)
+        self.output_filters = output_filters
+        self.bias_initializer = bias_initializer
+        conv_layers = [
+            layers.Conv2D(
+                256,
+                3,
+                padding="same",
+                kernel_initializer=initializers.RandomNormal(0.0, 0.01),
```
Comment (qlzh727) [inline keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:42]: not sure why the kernel initializer is hardcoded here. Might need some comments here.

File: keras_cv/models/object_detection/retina_net/retina_net_inference_test.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        model.prediction_decoder,
+        model.anchor_generator,
+        model.label_encoder,
+        model.classification_head,
+        model.box_head,
+    ]
+
+
+def _create_retina_nets(x, y, epochs=1):
+    retina_net = keras_cv.models.RetinaNet(
```
Comment (qlzh727) [inline keras_cv/models/object_detection/retina_net/retina_net_inference_test.py:112]: It took me some time to notice the difference between this model and the new_retina_net (which doesn't have the init weights). How about call this model to be pretrained_retina_net?

File: keras_cv/models/object_detection/retina_net/retina_net_inference_test.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            ),
+            keras_cv.metrics.COCORecall(
+                class_ids=range(20),
+                bounding_box_format="xyxy",
+                name="Recall",
+            ),
+        ],
+    )
+    retina_net.build((None, None, None, 3))
+    if epochs != 0:
```
Comment (qlzh727) [inline keras_cv/models/object_detection/retina_net/retina_net_inference_test.py:140]: Might need some comment about why we need to train this model here (and not the model down below.)

File: keras_cv/models/object_detection/retina_net/retina_net_inference_test.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            ):
+                self.assertAllEqual(weight, weight_new)
+
+    @pytest.mark.skipif(os.name == "nt", reason="tempfile does not work on windows")
+    def test_weight_loading(self):
+        x, y = _create_bounding_box_dataset(bounding_box_format="xywh")
+        retina_net, new_retina_net = _create_retina_nets(x, y, epochs=1)
+
+        tmp = tempfile.mkdtemp()
+        retina_net.save_weights(f"{tmp}/checkpoint.h5")
```
Comment (qlzh727) [inline keras_cv/models/object_detection/retina_net/retina_net_inference_test.py:59]: Can you also verify the tf format, which is probably the major use case?

https://www.tensorflow.org/api_docs/python/tf/keras/Model#load_weights

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:21]: it is actually quite weird that the training worked for this.  Even worse, training doesn't work AFTER this change... not sure what is going on there.

File: keras_cv/models/object_detection/retina_net/retina_net_inference_test.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (qlzh727) [inline keras_cv/models/object_detection/retina_net/retina_net_inference_test.py:1]: Any reason to have this as a standalone test file?

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: keras_cv/models/object_detection/retina_net/retina_net_inference_test.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net_inference_test.py:1]: Basically just QOL for me, there are a lot of RetinaNet tests.  an additional aspect is timeouts

File: keras_cv/models/object_detection/retina_net/retina_net_inference_test.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net_inference_test.py]: done

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:42]: Added a comment

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py [hunk 7]
```diff
@@ -30,7 +30,7 @@ class PredictionHead(layers.Layer):
         or the box regression head depending on `output_filters`.
     """
 
-    def __init__(self, output_filters, bias_initializer, **kwargs):
+    def __init__(self, output_filters, bias_initializer, conv_layers=1, **kwargs):
```
Comment (qlzh727) [inline keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py:33]: Note that the default value is changed from 4 to 1, so this might affect some existing prod code (if any)

File: keras_cv/models/object_detection/retina_net/retina_net_inference_test.py [hunk 8]
```diff
@@ -30,46 +30,79 @@ def cleanup_global_session(self):
...
+            pretrained_decoder.suppression_layer.iou_threshold,
+        )
+
+    def test_savedmodel_creation(self):
+        x, y = _create_bounding_box_dataset(bounding_box_format="xywh")
+        pretrained_retina_net, new_retina_net = _create_retina_nets(x, y, epochs=1)
+
+        tmp = tempfile.mkdtemp()
+        pretrained_retina_net.save(f"{tmp}/checkpoint/")
+        load_model = tf.saved_model.load(f"{tmp}/checkpoint/")
```
Comment (qlzh727) [inline keras_cv/models/object_detection/retina_net/retina_net_inference_test.py:80]: I think you can either use this, or `new_retina_net.load_weights(f"{tmp}/checkpoint/")`.

For the returned result from model.load_weights(), you can assert all the weights have consumed, eg assert_consumed().

https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint#restore

File: keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py (hunk 7 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py]: there is no prod code, but there are actually greater issues here 😨 the integration tests fail if you set `prediction_head.trainable=False`, but if trainable they fail.

**Comment by LukeWood** [comment]
@fchollet I'm going to merge the weight loading PR, then follow up by fixing the gradient based bugs.
